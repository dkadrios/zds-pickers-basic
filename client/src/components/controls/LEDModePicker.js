import React, { Component } from 'react'
import PropTypes from 'prop-types'
import IconButton from '@material-ui/core/IconButton'
import ArrowDropDown from '@material-ui/icons/ArrowDropDown'
import {
  ACTIVITY_LED_MODE_ALWAYS_OFF,
  ACTIVITY_LED_MODE_ALWAYS_ON,
  ACTIVITY_LED_MODE_NORMALLY_OFF,
  ACTIVITY_LED_MODE_NORMALLY_ON,
} from '../../midi/sysex'

const options = [
  { value: ACTIVITY_LED_MODE_NORMALLY_ON, label: 'Flicker on activity; Normally ON' },
  { value: ACTIVITY_LED_MODE_NORMALLY_OFF, label: 'Flicker on activity; Normally OFF' },
  { value: ACTIVITY_LED_MODE_ALWAYS_ON, label: 'Always ON' },
  { value: ACTIVITY_LED_MODE_ALWAYS_OFF, label: 'Always OFF' },
]

class LEDModePicker extends Component {
  static propTypes = {
    selectedValue: PropTypes.number.isRequired,
    setFlags: PropTypes.func.isRequired,
    serialMidiOutEnabled: PropTypes.bool.isRequired,
    usbMidiOutEnabled: PropTypes.bool.isRequired,
    children: PropTypes.node,
  }

  static defaultProps = { children: null }

  state = {
    anchorEl: null,
  }

  handleClick = (event) => {
    this.setState({ anchorEl: event.currentTarget })
  }

  handleMenuItemClick = (value) => {
    const { serialMidiOutEnabled, setFlags, usbMidiOutEnabled } = this.props
    setFlags(value, serialMidiOutEnabled, usbMidiOutEnabled)
    this.handleClose()
  }

  handleClose = () => {
    this.setState({ anchorEl: null })
  }

  render() {
    const { anchorEl } = this.state
    const { children, selectedValue } = this.props

    return (
      <>
        <IconButton
          aria-haspopup="true"
          aria-owns={anchorEl ? 'simple-menu' : null}
          onClick={this.handleClick}
        >
          <Tooltip
            enterDelay={250}
            placement="top"
            title="LED mode"
          >
            <div style={{ display: 'flex', flexFlow: 'row nowrap' }}>
              {children}
              <ArrowDropDown />
            </div>
          </Tooltip>
        </IconButton>

        <Menu
          anchorEl={anchorEl}
          id="simple-menu"
          onClose={this.handleClose}
          open={Boolean(anchorEl)}
        >
          {options.map(({ value, label }, index) => (
            <MenuItem
              key={value}
              onClick={() => this.handleMenuItemClick(value)}
              selected={index === selectedValue}
              tag={`miValue${value}`}
            >
              {label}
            </MenuItem>
          ))}
        </Menu>
      </>
    )
  }
}

export default LEDModePicker
