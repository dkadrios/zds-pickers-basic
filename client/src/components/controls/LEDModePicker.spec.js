import React from 'react'
import LEDModePicker from './LEDModePicker'

const setFlags = jest.fn()

const props = {
  selectedValue: 0,
  setFlags,
  disabled: false,
  serialMidiOutEnabled: true,
  usbMidiOutEnabled: true,
}

describe('LEDModePicker tests', () => {
  beforeEach(jest.resetAllMocks)

  it('renders correctly', () => {
    shallowExpect(LEDModePicker)(props).toMatchSnapshot()
  })

  it('clicking should open the menu', () => {
    const wrapper = shallow(<LEDModePicker {...props} />)
    wrapper.instance().handleClick({ currentTarget: 'target' })
    expect(wrapper.state('anchorEl')).toEqual('target')
  })

  it('clicking a menuitem should change state', () => {
    const wrapper = shallow(<LEDModePicker {...props} />)
    wrapper.instance().handleMenuItemClick(2)
    wrapper
      .find('WithStyles(MenuItem)')
      .filter('[tag="miValue3"]')
      .simulate('click')
    expect(setFlags).toHaveBeenCalledWith(3, true, true)
  })
})
