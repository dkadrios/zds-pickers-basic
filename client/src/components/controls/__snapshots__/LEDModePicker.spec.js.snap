// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LEDModePicker tests renders correctly 1`] = `
<Fragment>
  <WithStyles(IconButton)
    aria-haspopup="true"
    aria-owns={null}
    onClick={[Function]}
  >
    <WithStyles(Tooltip)
      enterDelay={250}
      placement="top"
      title="LED mode"
    >
      <div
        style={
          Object {
            "display": "flex",
            "flexFlow": "row nowrap",
          }
        }
      >
        <pure(ArrowDropDownIcon) />
      </div>
    </WithStyles(Tooltip)>
  </WithStyles(IconButton)>
  <WithStyles(Menu)
    anchorEl={null}
    id="simple-menu"
    onClose={[Function]}
    open={false}
  >
    <WithStyles(MenuItem)
      key="0"
      onClick={[Function]}
      selected={true}
      tag="miValue0"
    >
      Flicker on activity; Normally ON
    </WithStyles(MenuItem)>
    <WithStyles(MenuItem)
      key="1"
      onClick={[Function]}
      selected={false}
      tag="miValue1"
    >
      Flicker on activity; Normally OFF
    </WithStyles(MenuItem)>
    <WithStyles(MenuItem)
      key="2"
      onClick={[Function]}
      selected={false}
      tag="miValue2"
    >
      Always ON
    </WithStyles(MenuItem)>
    <WithStyles(MenuItem)
      key="3"
      onClick={[Function]}
      selected={false}
      tag="miValue3"
    >
      Always OFF
    </WithStyles(MenuItem)>
  </WithStyles(Menu)>
</Fragment>
`;
